tcpdump.c

```diff
@@ -0,0 +1,239 @@
+
+    CHECK_FUNC(compile_bpf(MAX_PKT_LEN - 1, link_type, &bpf_filter, bpf_str, opt));
+
+    // for debug purpose
+    // the 2 stands for how i print the bpf 
+    dump_bpf(&bpf_filter, 2); 
+    CHECK(setsockopt(sock, SOL_SOCKET, SO_ATTACH_FILTER, &bpf_filter, sizeof(bpf_filter)) >= 0);
+
+cleanup:
+    free(bpf_buffer); // Best effort.
+    return ret_status;
+}
+
+static error_status_t iface_name_to_index(int sock, const char* if_name, int* if_index) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    struct ifreq ifr = {0};
+    size_t if_name_len = 0;
+
+    CHECK(if_name != NULL);
+    CHECK(if_index != NULL);
+    
+    if_name_len = strlen(if_name);
+
+    CHECK_STR(if_name_len < sizeof(ifr.ifr_name), "Interface name too long");
+    memcpy(ifr.ifr_name, if_name, if_name_len);
+    // ioctl for mapping if_name to if_index
+    CHECK(ioctl(sock, SIOCGIFINDEX, &ifr) != -1);
+
+    *if_index = ifr.ifr_ifindex;
+
+cleanup:
+    return ret_status;
+}
+
+static error_status_t iface_set_promisc(int sock, const char* if_name, bool set) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int if_index = -1;
+    int opt_name = -1;
+    struct packet_mreq mreq = {0};
+    
+    CHECK(if_name != NULL);
+    CHECK_FUNC(iface_name_to_index(sock, if_name, &if_index));
+
+    mreq.mr_ifindex = if_index;
+    mreq.mr_type = PACKET_MR_PROMISC;
+
+	opt_name = set ? PACKET_ADD_MEMBERSHIP : PACKET_DROP_MEMBERSHIP;
+    CHECK(setsockopt(sock, SOL_PACKET, opt_name, &mreq, sizeof(mreq)) != -1);
+    // TODO: this is a side effect. what if i don't want to print?
+    printf("set interface: %s to promisc %d\n", if_name, set);
+
+cleanup:
+    return ret_status;
+}
+
+static error_status_t bind_raw_socket(int sock, const char* if_name) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int if_index = -1;
+    struct sockaddr_ll sll = {0};
+    
+    CHECK(if_name != NULL);
+    CHECK_FUNC(iface_name_to_index(sock, if_name, &if_index));
+
+    // to bind a raw socket to a specific interface 
+    sll.sll_family = AF_PACKET;
+    sll.sll_protocol = htons(ETH_P_ALL);
+    sll.sll_ifindex = if_index;
+
+    CHECK_STR(bind(sock, (struct sockaddr*)&sll, sizeof(sll)) >= 0, "Failed binding to interface");
+
+cleanup:
+    // TODO: i don't close here even if it fails because i want main func to close it? 
+    //close(sock);
+    return ret_status;
+
+}
+
+static error_status_t create_raw_socket(int* res) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int sock = -1;
+    
+    CHECK(res != NULL);
+
+    sock = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+    CHECK_STR(sock >= 0, "Creating raw socket failed");
+
+    *res = sock;
+cleanup:
+    return ret_status;
+}
+static error_status_t handle_packet(unsigned char* packet) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    
+    CHECK(packet != NULL);
+
+    // Basic packet parser
+    // add more: ip level, tcp/udp, blah blah
+    struct ethhdr* eth_h = (struct ethhdr*)packet;
```

Comment:
What if the packet is too small for this to make sense?
handle_packet should get the packet size as a parameter, and refer to it where necessary